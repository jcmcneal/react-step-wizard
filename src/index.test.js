import React from 'react';
import Renderer from 'react-test-renderer';
// import Enzyme from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';

import StepWizard from './index';

// Enzyme.configure({ adapter: new Adapter() });

/** Helper Functions */
const Step1 = () => <div>Step 1</div>;
const Step2 = () => <div>Step 2</div>;
const Step3 = () => <div>Step 3</div>;
const Step4 = () => <div>Step 4</div>;

const render = jsx => Renderer.create(jsx);
const getInstance = (component) => {
    const wrapper = render(component);
    return wrapper.getInstance();
};
const getState = component => getInstance(component).state;
const takeSnapshot = test => expect(test).toMatchSnapshot();
const getTreeSnapshot = component => takeSnapshot(render(component));
const testComponent = (component) => {
    getTreeSnapshot(component);

    const state = getState(component);
    takeSnapshot(state);

    return state;
};
const basicComponent = () => (
    getInstance((
        <StepWizard>
            <Step1 />
            <Step2 />
            <Step3 />
        </StepWizard>
    ))
);

/** Capture console errors */
console.error = jest.fn();

/** Unit Tests */
describe('Step Wizard Component', () => {
    it('empty component', () => {
        testComponent(<StepWizard />);
    });

    it('html element steps', () => {
        testComponent((
            <StepWizard>
                <Step1 />>
            </StepWizard>
        ));
    });

    it('with 3 basic components', () => {
        testComponent((
            <StepWizard>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
    });

    it('initialStep set to 2', () => {
        testComponent((
            <StepWizard initialStep={2}>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
    });

    it('isHashEnabled - true', () => {
        testComponent((
            <StepWizard isHashEnabled>
                <Step1 hashKey={'First'} />
                <Step2 />
                <Step3 hashKey={'The End!'} />
            </StepWizard>
        ));
    });

    it('isHashEnabled - false', () => {
        testComponent((
            <StepWizard>
                <Step1 hashKey={'First'} />
                <Step2 />
                <Step3 hashKey={'The End!'} />
            </StepWizard>
        ));
    });

    it('isLazyMount - true', () => {
        testComponent((
            <StepWizard isLazyMount initialStep={2}>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
    });

    it('nav', () => {
        testComponent((
            <StepWizard nav={<div>Nav</div>}>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
    });

    it('instance', () => {
        testComponent((
            <StepWizard instance={SW => SW}>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
    });

    it('conditional step when not visible', () => {
        testComponent((
        <StepWizard>
            <Step1 />
            <Step2 />
            <Step3 />
            { false && <Step4 />}
        </StepWizard>))
    })

    it('conditional step when visible', () => {
        testComponent((
        <StepWizard>
            <Step1 />
            <Step2 />
            <Step3 />
            { true && <Step4 />}
        </StepWizard>))
    })

    it('garbage props', () => {
        console.error.mockClear();

        testComponent((
            <StepWizard
                initialStep={true}
                isHashEnabled={'fdjlk'}
                onStepChange={true}
                transitions={false}
            >
                <Step1 hashKey={'First *3@ #$~!,";'} />
                <Step2 />
                <Step3 hashKey={1} />
            </StepWizard>
        ));

        expect(console.error).toHaveBeenCalledWith('Warning: Failed prop type: Invalid prop `initialStep` of type `boolean` supplied to `StepWizard`, expected `number`.\n    in StepWizard');
        expect(console.error).toHaveBeenCalledWith('Warning: Failed prop type: Invalid prop `isHashEnabled` of type `string` supplied to `StepWizard`, expected `boolean`.\n    in StepWizard');
        expect(console.error).toHaveBeenCalledWith('Warning: Failed prop type: Invalid prop `onStepChange` of type `boolean` supplied to `StepWizard`, expected `function`.\n    in StepWizard');
        expect(console.error).toHaveBeenCalledWith('Warning: Failed prop type: Invalid prop `transitions` of type `boolean` supplied to `StepWizard`, expected `object`.\n    in StepWizard');
    });
});

describe('Step Wizard Functions', () => {
    it('lastStep', () => {
        const wrapper = basicComponent();
        wrapper.lastStep();
        takeSnapshot(wrapper.state);
        expect(wrapper.state.activeStep).toEqual(2);
    });
    it('firstStep', () => {
        const wrapper = basicComponent();
        wrapper.firstStep();
        takeSnapshot(wrapper.state);
        expect(wrapper.state.activeStep).toEqual(0);
    });
    it('nextStep', () => {
        const wrapper = basicComponent();
        wrapper.nextStep();
        takeSnapshot(wrapper.state);
        expect(wrapper.state.activeStep).toEqual(1);
    });
    it('previousStep', () => {
        const wrapper = basicComponent();
        wrapper.lastStep();
        wrapper.previousStep();
        takeSnapshot(wrapper.state);
        expect(wrapper.state.activeStep).toEqual(1);
    });
    it('goToStep', () => {
        const wrapper = basicComponent();
        wrapper.goToStep(3);
        takeSnapshot(wrapper.state);
        expect(wrapper.state.activeStep).toEqual(2);
    });
    it('isInvalidStep', () => {
        const wrapper = basicComponent();
        expect(wrapper.isInvalidStep(-1)).toEqual(true);
        expect(wrapper.isInvalidStep(0)).toEqual(false);
        expect(wrapper.isInvalidStep(1)).toEqual(false);
        expect(wrapper.isInvalidStep(2)).toEqual(false);
        expect(wrapper.isInvalidStep(3)).toEqual(true);
        expect(wrapper.isInvalidStep(50)).toEqual(true);
    });
    it('onStepChange', () => {
        const onStepChange = jest.fn();
        const wrapper = getInstance((
            <StepWizard onStepChange={onStepChange}>
                <Step1 />
                <Step2 />
                <Step3 />
            </StepWizard>
        ));
        wrapper.lastStep();
        wrapper.previousStep();
        wrapper.firstStep();
        wrapper.nextStep();
        wrapper.goToStep(1);
        expect(onStepChange).toHaveBeenCalledTimes(5);

        wrapper.firstStep(); // no change
        expect(onStepChange).toHaveBeenCalledTimes(5);
        takeSnapshot(wrapper.state);
    });
    // getHash
    // initialState
    // isInvalidStep
    // onHashChange
    // setActiveStep
    // updateHash
});
