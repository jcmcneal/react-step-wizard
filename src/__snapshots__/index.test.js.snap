// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step Wizard Component conditional step 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 4
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component conditional step 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "3": "step4",
    "step1": 0,
    "step2": 1,
    "step3": 2,
    "step4": 3,
  },
}
`;

exports[`Step Wizard Component conditional step when not visible 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component conditional step when not visible 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Component conditional step when visible 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 4
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component conditional step when visible 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "3": "step4",
    "step1": 0,
    "step2": 1,
    "step3": 2,
    "step4": 3,
  },
}
`;

exports[`Step Wizard Component empty component 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  />
</div>
`;

exports[`Step Wizard Component empty component 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {},
}
`;

exports[`Step Wizard Component garbage props 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component garbage props 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "First *3@ #$~!,\\";",
    "1": 2,
    "2": 1,
    "First *3@ #$~!,\\";": 0,
    "step2": 1,
  },
}
`;

exports[`Step Wizard Component html element steps 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      &gt;
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component html element steps 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "step1": 0,
    "step2": 1,
  },
}
`;

exports[`Step Wizard Component initialStep set to 2 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step  active"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component initialStep set to 2 2`] = `
Object {
  "activeStep": 1,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Component instance 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component instance 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Component isHashEnabled - false 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component isHashEnabled - false 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "First",
    "1": "step2",
    "2": "The End!",
    "First": 0,
    "The End!": 2,
    "step2": 1,
  },
}
`;

exports[`Step Wizard Component isHashEnabled - true 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component isHashEnabled - true 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "First",
    "1": "step2",
    "2": "The End!",
    "First": 0,
    "The End!": 2,
    "step2": 1,
  },
}
`;

exports[`Step Wizard Component isLazyMount - true 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 2
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component isLazyMount - true 2`] = `
Object {
  "activeStep": 1,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Component nav 1`] = `
<div
  className={null}
>
  <div
    currentStep={1}
    firstStep={[Function]}
    goToStep={[Function]}
    isActive={false}
    lastStep={[Function]}
    nextStep={[Function]}
    previousStep={[Function]}
    totalSteps={3}
  >
    Nav
  </div>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component nav 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Component with 3 basic components 1`] = `
<div
  className={null}
>
  <div
    className="step-wrapper"
  >
    <div
      className="step  active"
    >
      <div>
        Step 1
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 2
      </div>
    </div>
    <div
      className="step"
    >
      <div>
        Step 3
      </div>
    </div>
  </div>
</div>
`;

exports[`Step Wizard Component with 3 basic components 2`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions firstStep 1`] = `
Object {
  "activeStep": 0,
  "classes": Object {},
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions goToStep 1`] = `
Object {
  "activeStep": 2,
  "classes": Object {
    "0": "animated fadeOutLeft",
    "2": "animated fadeInRight",
  },
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions lastStep 1`] = `
Object {
  "activeStep": 2,
  "classes": Object {
    "0": "animated fadeOutLeft",
    "2": "animated fadeInRight",
  },
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions nextStep 1`] = `
Object {
  "activeStep": 1,
  "classes": Object {
    "0": "animated fadeOutLeft",
    "1": "animated fadeInRight",
  },
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions onStepChange 1`] = `
Object {
  "activeStep": 0,
  "classes": Object {
    "0": "animated fadeInLeft",
    "1": "animated fadeOutRight",
    "2": "animated fadeOutRight",
  },
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;

exports[`Step Wizard Functions previousStep 1`] = `
Object {
  "activeStep": 1,
  "classes": Object {
    "0": "animated fadeOutLeft",
    "1": "animated fadeInLeft",
    "2": "animated fadeOutRight",
  },
  "hashKeys": Object {
    "0": "step1",
    "1": "step2",
    "2": "step3",
    "step1": 0,
    "step2": 1,
    "step3": 2,
  },
}
`;
